proxy/src/control/discovery.rs

```diff
@@ -199,27 +206,76 @@ where
     type DiscoverError = ();
 
     fn poll(&mut self) -> Poll<Change<Self::Key, Self::Service>, Self::DiscoverError> {
-        let up = self.rx.poll();
-        trace!("watch: {:?}", up);
-        let update = match up {
-            Ok(Async::Ready(Some(update))) => update,
-            Ok(Async::Ready(None)) => unreachable!(),
-            Ok(Async::NotReady) => return Ok(Async::NotReady),
-            Err(_) => return Err(()),
-        };
-
-        match update {
-            | Update::Insert(addr, meta) => {
-                let service = self.bind.bind(&addr)
-                    .map(|svc| meta.label(svc))
-                    .map_err(|_| ())?;
+        loop {
+            let up = self.rx.poll();
+            trace!("watch: {:?}", up);
+            let update = match up {
+                Ok(Async::Ready(Some(update))) => update,
+                Ok(Async::Ready(None)) => unreachable!(),
+                Ok(Async::NotReady) => return Ok(Async::NotReady),
+                Err(_) => return Err(()),
+            };
 
-                Ok(Async::Ready(Change::Insert(addr, service)))
-            },
-            // TODO: handle metadata changes by changing the labeling
-            // middleware to hold a `futures-watch::Watch` on the label value,
-            // so it can be updated.
-            Update::Remove(addr) => Ok(Async::Ready(Change::Remove(addr))),
+            match update {
+                Update::Insert(addr, meta) => {
+                    // Construct a watch for the `Labeled` middleware that will
+                    // wrap the bound service, and insert the store into our map
+                    // so it can be updated later.
+                    let (watch_labels, update_labels) =
+                        futures_watch::Watch::new(meta.metric_labels);
+                    self.metric_labels.insert(addr, update_labels);
+
+                    let service = self.bind.bind(&addr)
+                        .map(|svc| Labeled::new(svc, watch_labels))
+                        .map_err(|_| ())?;
+
+                    return Ok(Async::Ready(Change::Insert(addr, service)))
+                },
+                Update::ChangeMetadata(addr, meta) => match self
```

Comment:
can you please wrap this sub-match in braces (+newline+indent)? this is a little too subtle for me.

Or perhaps this all can be moved into its own method?
```rustlang
fn update_metadata(&mut self, addr: SocketAddr, meta: ..) -> Result<(), ...>
```

If that's possible and not too hard, strongly encouraged!

Reply:
Done in 53c2eb9204e67df1dbf5adbb2ab47436d937f74a --- don't know why I didn't factor that out earlier. :)